datalad/support/annexrepo.py

```diff
@@ -332,6 +365,142 @@ def annex_add_to_git(self, files):
+            and only a single string will work here.
+            TODO: figure out, how to document this behaviour
+                  properly everywhere
+
+        url: str
+
+        options: list
+            options to the annex command
+        """
+
+        annex_options = ['--file=%s' % file[0]] + options + [url]
+        self._run_annex_command('addurl', annex_options=annex_options)
+
+    def annex_addurls(self, urls, options=[]):
+        """Downloads each url to its own file, which is added to the annex.
+
+        Parameters:
+        -----------
+        urls: list
+
+        options: list
+            options to the annex command
+        """
+
+        self._run_annex_command('addurl', annex_options=options + urls)
+
+    @normalize_paths
+    def annex_rmurl(self, file, url):
+        """Record that the file is no longer available at the url.
+
+        Parameters:
+        -----------
+        file: str
+
+        url: str
+        """
+
+        self._run_annex_command('rmurl', annex_options=[file[0]] + [url])
+
+    @normalize_paths
+    def annex_drop(self, files):
+        """Drops the content of annexed files from this repository.
+
+        Drops only if possible with respect to required minimal number of
+        available copies.
+
+        Parameters:
+        -----------
+        files: list
+        """
+        # TODO: options needed
+
+        self._run_annex_command('drop', annex_options=files)
+
+    @normalize_paths
+    def annex_whereis(self, files):
+        """Lists repositories that have actual content of file(s).
+
+        Parameters:
+        -----------
+        files: list
+            files to look for
+
+        Returns:
+        --------
+        {file: [desc]}
+            where `file` is every file from `files`,
+            'whereis' was successfully ran on.
+            [desc] contains a unicode describing the remote for each remote,
+            which was found by git-annex whereis, like:
+
+            u'me@mycomputer:~/where/my/repo/is [origin]' or
+            u'web' or
+            u'me@mycomputer:~/some/other/clone'
+        """
+
+        try:
+            out, err = self._run_annex_command('whereis',
+                                              annex_options=['--json'] + files)
+        except CommandError, e:
+            # if multiple files, whereis may technically fail,
+            # but still returns correct response
+            if e.code == 1 and e.stdout.startswith('{'):
+                out = e.stdout
+            else:
+                raise e
+
+        json_objects = []
+        for line in out.split(linesep):
+            if line.startswith('{'):
+                json_objects.append(json.loads(line))
+
+        result = {}
+        for item in json_objects:
+            if item.get('success'):
+                result[item.get('file')] = \
+                  [remote.get('description') for remote in item.get('whereis')]
```

Comment:
You think THAT is better to read? :)

2015-04-15 19:39 GMT+02:00 Yaroslav Halchenko <notifications@github.com>:

> In datalad/support/annexrepo.py
> <https://github.com/yarikoptic/datalad/pull/2#discussion_r28443206>:
>
> > +            # but still returns correct response
> > +            if e.code == 1 and e.stdout.startswith('{'):
> > +                out = e.stdout
> > +            else:
> > +                raise e
> > +
> > +        json_objects = []
> > +        for line in out.split(linesep):
> > +            if line.startswith('{'):
> > +                json_objects.append(json.loads(line))
> > +
> > +        result = {}
> > +        for item in json_objects:
> > +            if item.get('success'):
> > +                result[item.get('file')] = \
> > +                  [remote.get('description') for remote in item.get('whereis')]
>
> oh sh.t
>
> result = {item.get('file'): [remote.get('description') for remote in item.get('whereis')]
>           for item in json_objects
>           if item.get('success')}
>
> ? ;-)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/yarikoptic/datalad/pull/2/files#r28443206>.
>